src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ComponentRenderingFunctionalTests.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using AngleSharp.Parser.Html;
+using BasicWebSite.Services;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.FunctionalTests
+{
+    public class ComponentRenderingFunctionalTests : IClassFixture<MvcTestFixture<BasicWebSite.Startup>>
+    {
+        public ComponentRenderingFunctionalTests(MvcTestFixture<BasicWebSite.Startup> fixture)
+        {
+            Client = fixture.WithWebHostBuilder(b => b.ConfigureServices(ConfigureServices)).CreateClient();
+            LoopHandler.InnerHandler = fixture.Server.CreateHandler();
+
+            void ConfigureServices(IServiceCollection services) =>
+                services.AddSingleton(new WeatherForecastService(new HttpClient(LoopHandler)
+                {
+                    BaseAddress = fixture.ClientOptions.BaseAddress
+                }));
+        }
+
+        private class LoopHttpHandler : DelegatingHandler
+        {
+        }
```

Comment:
Its an indirection level to create a loopback connection against the test server, so that you can have a service configured that calls back into test server from within the app

Reply:
I don't really get it from looking at the code. Can this be explained with some comments? TBH I'm not totally certain why these tests are so sophisticated

Reply:
I added some comments. And cleaned up the code.

This test just shows that we can prerrender an async action like the ones in the Components template.